Microsoft.Azure.Cosmos/src/Resource/Container/CosmosContainerCore.cs

```
@@ -200,12 +204,37 @@ internal Task<string> GetRID(CancellationToken cancellationToken)
                             .ContinueWith(containerSettingsTask => containerSettingsTask.Result?.ResourceId, cancellationToken);
         }
 
-        internal Task<PartitionKeyDefinition> GetPartitionKeyDefinitionAsync(CancellationToken cancellationToken = default(CancellationToken))
+        internal virtual Task<PartitionKeyDefinition> GetPartitionKeyDefinitionAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
             return this.GetCachedContainerSettingsAsync(cancellationToken)
                             .ContinueWith(containerSettingsTask => containerSettingsTask.Result?.PartitionKey, cancellationToken);
```

Should we be throwing if the container doesn't exist and we are trying to access the partition key value so we can avoid the current ambiguity of not knowing if the container settings is null or if the pk on the settings is null?

---

Sorry, this should be exception less.

---

Exception less is only for NW interaction. We can still throw Argument*Exception in-case of invalid inputs. 

For invalidate state's lets enumerate and walkthrough
- Non-existent container: should return 424 (Dependency failure)
- PK definition null (with new API version is not expected but as defense practice we can throw ArgumentException)

---

now throws on null pk definition. cache throws on non-existent collection